### PR TITLE
[Frontend] calling `qjit(keep_intermediate=True)` multiple times

### LINKED ISSUE #291: Setting qjit keep_intermediate=True flag leads to memory corruptions
Changing `@qjit` to `@qjit(keep_intermediate=True)` in some tests leads to memory corruptions (seen as mismatched test results).

This behavior was observed in [test_adjoint.py](https://github.com/PennyLaneAI/catalyst/blob/422dc14879071039a9a64eb781627b6b441174de/frontend/test/pytest/test_adjoint.py#L37), [test_gradient.py](https://github.com/PennyLaneAI/catalyst/blob/422dc14879071039a9a64eb781627b6b441174de/frontend/test/pytest/test_gradient.py#L155) and in some others.

The affected Catalyst revision is 422dc148

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Having a single workspace for a single python function to be compiled could lead to an error when the same function was compiled multiple times or another function with the same name was compiled. It also wasn't entirely clear what the behaviour should be when using `keep_intermediate` and a function was recompiled.

**Description of the Change:** Now when two functions of the request the same name, a new directory will be created appending `_$NUM`. 

**Benefits:** No issues with compiling functions with the same name.

**Possible Drawbacks:**

**Related GitHub Issues:** closes #291 

[sc-47499]

### CODE REVIEW

**Comment by dime10** [comment]
Maybe that's a silly question, but why wouldn't the compiler driver not just overwrite any existing files in the working directory already exists? I believe that was the previous behaviour in Python.

**Comment by erick-xanadu** [comment]
Not a silly question! I'm not sure, let me dig into it.

File: frontend/catalyst/compiler.py [hunk 1]
```diff
@@ -407,23 +397,30 @@ def run(self, mlir_module, *args, **kwargs):
             **kwargs,
         )
 
-    def get_output_of(self, pipeline) -> Optional[str]:
+    def get_output_of(self, workspace, pipeline) -> Optional[str]:
```
Comment (dime10) [inline frontend/catalyst/compiler.py:400]: Is workspace unused in this function (besides the checks)?

File: frontend/catalyst/compilation_pipelines.py [hunk 2]
```diff
@@ -444,6 +448,97 @@ def __call__(self, *args, **kwargs):
         return result
 
 
+class Directory:
```
Comment (dime10) [inline frontend/catalyst/compilation_pipelines.py:451]: Should these classes (`Directory`, `WorkspaceManager`) go into utils? That might avoid the `too-many-lines` flag, and also the circular import in the other file.

**Review by dime10 — COMMENTED** [review]
Also a good solution!

I wonder what the trade-off is for handling this in the frontend though, if it is an issue in the compiler driver. With this design, would it simply be an "API misuse" if the compiler driver was fed a directory that already had files in it?

File: frontend/catalyst/compiler.py (hunk 1 above)
Comment (erick-xanadu) [reply frontend/catalyst/compiler.py:400]: Ah yes, I think this was an attempt to make the compiler functional (i.e., without any state), but I forgot to remove `last_compiler_output`. I'll work on it tomorrow.

File: frontend/catalyst/compilation_pipelines.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/catalyst/compilation_pipelines.py]: https://github.com/PennyLaneAI/catalyst/pull/306/commits/e4fd4f7dd28d97a082756d389feda2ffeb577e72 Thanks!

**Review by dime10 — APPROVED** [review]
🦝

**Pull request merged**